source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -0,0 +1,237 @@
+#include "source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.h"
+
+#include "envoy/extensions/network/dns_resolver/getaddrinfo/v3/getaddrinfo_dns_resolver.pb.h"
+#include "envoy/network/dns_resolver.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+// This resolver uses getaddrinfo() on a dedicated resolution thread. Thus, it is only suitable
+// currently for relatively low rate resolutions. In the future a thread pool could be added if
+// desired.
+class GetAddrInfoDnsResolver : public DnsResolver, public Logger::Loggable<Logger::Id::dns> {
+public:
+  GetAddrInfoDnsResolver(Event::Dispatcher& dispatcher, Api::Api& api)
+      : dispatcher_(dispatcher),
+        resolver_thread_(api.threadFactory().createThread([this] { resolveThreadRoutine(); })) {}
+
+  ~GetAddrInfoDnsResolver() override {
+    {
+      absl::MutexLock guard(&mutex_);
+      shutting_down_ = true;
+    }
+
+    resolver_thread_->join();
+  }
+
+  // DnsResolver
+  ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                          ResolveCb callback) override {
+    ENVOY_LOG(debug, "adding new query [{}] to pending queries", dns_name);
+    auto new_query = std::make_unique<PendingQuery>(dns_name, dns_lookup_family, callback);
+    absl::MutexLock guard(&mutex_);
+    pending_queries_.emplace_back(std::move(new_query));
+    return pending_queries_.back().get();
+  }
+
+  void resetNetworking() override {}
+
+private:
+  class PendingQuery : public ActiveDnsQuery {
+  public:
+    PendingQuery(const std::string& dns_name, DnsLookupFamily dns_lookup_family, ResolveCb callback)
+        : dns_name_(dns_name), dns_lookup_family_(dns_lookup_family), callback_(callback) {}
+
+    void cancel(CancelReason) override {
+      ENVOY_LOG(debug, "cancelling query [{}]", dns_name_);
+      cancelled_ = true;
+    }
+
+    const std::string dns_name_;
+    const DnsLookupFamily dns_lookup_family_;
+    ResolveCb callback_;
+    bool cancelled_{false};
+  };
+  // Must be a shared_ptr for passing around via post.
+  using PendingQuerySharedPtr = std::shared_ptr<PendingQuery>;
+
+  // RAII wrapper to free the `addrinfo`.
+  class AddrInfoWrapper : NonCopyable {
+  public:
+    AddrInfoWrapper(addrinfo* info) : info_(info) {}
+    ~AddrInfoWrapper() { Api::OsSysCallsSingleton::get().freeaddrinfo(info_); }
```

Comment:
It seems like unsafe to call `OsSysCallsImpl::freeaddrinfo` with nullptr (it will get nullptr when `OsSysCallsImpl::getaddrinfo` failed at https://github.com/envoyproxy/envoy/pull/22080/files#diff-01fdbaa5394bb19e54e31b01894d73299a3578fb2b54abcd8a5722b30db3ef80R175)

The man page didn't clear about this case https://man7.org/linux/man-pages/man3/getaddrinfo.3.html

But after googling, there is someone saw the segment fault when passing null to freeaddrinfo.

Also I'm thinking maybe we needn't pass the pointer to this wrapper, just create a class like:

```
Class AddrInfo : NonCopyable {
public:
    ~AddrInfo() { if (info_) {Api::OsSysCallsSingleton::get().freeaddrinfo(info_); } }

   addrinfo* info_{nullptr};
}
```

Then we pass the `AddrInfo::info_` to the `OsSysCallsImpl::getaddrinfo`. Or we ask `OsSysCallsImpl::getaddrinfo` accept a reference of `AddrInfo`, then no one will use `addrinfo*` nakely. But I don't if we want to add more logic for `OsSysCallsImpl::getaddrinfo` or just want to keep same with os interface.

Reply:
Thanks for the suggestion but I think I'm done making changes at this point. I test both the null and not null case and it seems to work ok. We can change it later if needed.

Reply:
cool, that works for me. This extension was added for android, and you tested that works well.